### ISSUE TITLE
Empty data type vs abstract data type

### ISSUE THREAD

**Issue by Ericson2314** [ghc-proposals/ghc-proposals#561]
```haskell
data Foo
```
means one thing in signatures (abstract data type) and another thing in regular modules (empty data type).

I think it is better for the former, and so we should come up with a new syntax for the latter. I am sure there is a GHC issue for this, but as it is a very user-facing bikeshed it might also belong here.

On idea that popped into my head is that if we allow leading and/or trailing `|` like
````haskell
data Bool = | True | False
data Bool = True | False |
```
(which could also help resolve some style guide bike-shedding)

then we could do
```haskell
data Void = |
```
to mean the empty data type.

-------

I was reminded of this because the [ChatGPT](https://openai.com/blog/chatgpt/) neural network kinda "knows" what Backpack is, but it utterly confused about how one writes an abstract data type in Haskell. When things that I suspect would confuse human students inded demonstrably confuse chat bots, I smirk, but also want to do something about it! We finally have (weak) evidence!

**Comment by int-index** [comment]
> I think it is better for the former, and so we should come up with a new syntax for the latter.

We already have one: `data Void where`, or `data Void where {}` if you will. Doesn't even require `-XGADTs`, only `-XGADTSyntax`.

**Comment by Ericson2314** [comment]
Ah yes, requiring the `where` is nice. I would be not be opposed to making `-XGADTSyntax` and `-XEmptyDataDecl` enforce that everywhere after a suitable deprecation period.

**Comment by int-index** [comment]
I'd support this proposal. Note, however, that there's a similar issue with classes. Is `class C a` an empty class (no methods) or an abstract class? This can also be disambiguated by an explicit `where`.

**Comment by goldfirere** [comment]
While I agree with the intent here, I would need a lot of convincing to make this change. It's not backward-compatible. And it will thus break code in the wild, but with little apparent upside.

This will break code in at least two ways:

1. Empty pattern matches.

Previously, a `data NoDef` could be used with `(\case {}) :: NoDef -> a` quite happily. Not so anymore.

2. Apartness

Previously, if we had `data NoDef1` and `data NoDef2`, we could have `instance C NoDef1` and `instance C NoDef2`, and these did not overlap. Not so anymore.

Though please argue your point if you like, the most I'd currently personally support in this direction is a warning (off by default but perhaps in `-Wcompat` and eventually `-Wall`) that suggests writing an explicit `where {}` to denote an empty definition.

And, @treeowl, I agree with you about data families. But I can't seem to go anywhere with that observation either. Maybe the closest is that one view of an ordinary `data` declaration is a `data family` and a universal `data instance` ("universal" = the arguments are just variables). I think this is a good way of thinking about datatypes. So abstract types are naturally the `data family` without the `data instance`.

**Comment by xplat** [comment]
Alternately, `data Foo` could be an empty datatype and `data Foo = _` could be an abstract one.  I don't love this or even like it because it makes the defined-as-null thing look like a non-definition and the non-defined thing look like a definition, but it wouldn't break existing code and it would make abstract types legitimately part of the vocabulary.

**Comment by Ericson2314** [comment]
> the most I'd currently personally support in this direction is a warning (off by default but perhaps in `-Wcompat` and eventually `-Wall`) that suggests writing an explicit `where {}` to denote an empty definition.

Right, and something like that is what I had in mind with "suitable deprecation period". I don't even mind if there is no concrete plan when the thing is eventually moved but it is in `-Wcompat`.

After I wrap up my current proposals, I have on for absurd patterns (https://github.com/ghc-proposals/ghc-proposals/discussions/423) I want to submit, and maybe after that would be a good time to think about this more seriously (basically, revisiting `-XEmptyData` after we revisited `-XEmptyCase`). But I thought I could at least open the issue right away.
